cap/doc.go

```diff
@@ -0,0 +1,279 @@
+
+The third argument (3) is an Opt, this one specifies how these tree nodes should get
+started. the LeftToRight value indicates that when starting, it should start
+with myWorker, then myOtherWorker and end with mySubsystem.
+
+This API will guarantee that the mySubsystem won't start until myOtherWorker
+starts, and that one won't start until myWorker starts. When root shuts down, it
+will stop each child node in reverse order, starting with mySubsystem, then
+myOtherWorker and finally myWorker.
+
+The fourth argument (4) is an Opt, this option specifies how children should get
+restart when on of them fails. In this particular example, only the failing
+child node will get restarted. Check the WithStrategy docs for more details.
+
+
+Once we have a SupervisorSpec, we have the blueprint to start a system, spawning
+each goroutine we need in order. Note all the wiring of resources is happening
+at build time, the allocation resources comes at later step.
+
+We will use this SupervisorSpec to create a Supervisor value.
+
+Supervisor
+
+To create Supervisor, you need to call the Start method of a SupervisorSpec
+value
+
+	sup, startErr := root.Start()
+
+This call will return a Supervisor record or an error. The error may happen when
+one of the children nodes is not able to allocate some IO resource and reports
+it failed to Start (see NewWithStart and NewSupervisorSpec for details).
+
+The Start call is synchronous, the program will not continue until all worker
+nodes and sub-trees are started in the specified order. For our SupervisorSpec
+above the following supervision tree will be spawned
+
+	root
+	|
+	` root/my-worker (spawned goroutine)
+	|
+	` root/my-other-worker (spawned goroutine)
+	|
+	 ` my-subsystem
+		|
+		` <other workers or sub-trees here>
+
+All your program should have allocated the resources and running as expected. If you want
+to stop the system, you can call the Stop method
+
+	stopErr := sup.Stop()
+
+The stopErr error will tell you if any of the children failed to stop. You may
+also join the main goroutine with the root supervisor monitor goroutine using
+the Wait method
+
+	supErr := sup.Wait()
+
+Resources in Workers
+
+There are times where an IO resource belongs to a single worker, and it doesn't
+make sense to consider a worker "started" until this resource is allocated.
+
+For this use case you can create a Worker goroutine that receives a callback
+function that indicates that the Worker was started (or failed to start).
+
+Lets use the NewWorkerWithNotifyStartFn function to model this behavior:
+
+	func newReportGenWorker(
+		name string,
+		dbConfig dbConfig,
+		reportCh chan Report,
+		opts ...cap.WorkerOpt,
+	) cap.Worker {
+		return cap.NewWorkerWithNotifyStartFn(
+			name,
+			func(ctx context.Cotnext, notifyStart cap.NotifyStartFn) error {
+				db, err := openDB(dbConfig)
+				if err != nil {
+					// (1)
+					notifyStart(err)
+					return err
+				}
+				// (2)
+				notifyStart(nil)
```

Comment:
a polymorphic `notifyStart` would be so nice. 🤷‍♂️

Reply:
Doing a polymorphic function here would ask me to receive a slice of errors. I don't like that, do you know of any other way